W-13714296 ACB AI feature for Pilot (#198)
mulesoft/docs-code-builder: modules/ROOT/pages/int-create-integrations-ai.adoc

= Creating Integrations Using Einstein

include::partial$acb-ide-logos.adoc[tags="both-ides"]

> The Generative Flow feature transforms your natural language prompt, which describes your business logic, into a flow that contains the connectors and components, in the order needed, to accomplish the specified task. 

Reviewer: Regarding "connectors and components," ACB calls connectors components, so maybe should be like below, "connectors and processors". ACB calls Mule components "processors" or "Core processors".

> For example, if you want to create a flow that reads contacts from Salesforce in JSON format, you can enter a prompt describing that task. The prompt is then transformed into a series of connectors and processors that are needed to execute that task, in this case: `Listener`, `Query`, `Transformer`, and `Logger`. A flow is generated, which you can then further customize to suit your needs. 

Reviewer: `Listener`, `Query`, `Transformer`, and `Logger`
We don't write these in code font. We'd use the name for them like you see in ACB, or their XML could be in code font, but we'd need the names of the XML elements

[[before-you-begin]]
== Before You Begin

Before you start creating your integration using Generative Flows:

* xref:start-acb.adoc[Set up and access the web or desktop IDE].

> * Ensure that you have the following permissions to use the Generative Flow feature:

Reviewer: Turns out they'll need other permissions to use ACB effectively. See the updated section on permissions: https://dev-docs-internal.kqa.msap.io/beta-code-builder/anypoint-code-builder/start-configure-permissions#permissions

[[access-einstein]]
== Access Einstein

There are two ways to open Einstein:

* When you create a new integration project, click the *Start with AI* card in the flow canvas. 
+
image::int-start-with-ai-card.png[Start with AI or Open Einstein card]

* Click the Einstein icon in the toolbar. 

image::int-einstein-icon-in-toolbar.png[Arrow pointing to Einstein icon in the toolbar]

[[create-integration-project]]
== Create a New Integration Project

To create a new integration project:

// Open the ACB IDE
include::partial$acb-reusable-steps.adoc[tags="open-ide"]
. From *Quick Actions*, click *Develop an Integration*:
+
image::develop-integration-option-mat.png["Develop an Integration link highlighted in the Getting started section"]

. Complete the *Develop an Integration* form.
+
[%header,cols="20a,60a"] 
|===
| Field Name | Field Value

| *Project Name* | Unique name for your project.

This name is used as the title and name of the integration specification file.
For example, if the project name is "My Integration", the name of the folder for the integration project is `my-integration`, and the configuration XML file name is `my-integration.xml`.

> | *Project Location* | Your home directory or another directory you create. Click *Browse* to select a directory for the integration if you don't want to use the default.

Reviewer: re: "if you don’t want to use the default."
I'm not sure about a default. I need to check into this further though.

Author: @dukesphere Maybe default isn't the right word, but that field is always prepopulated with a location.

Reviewer: It's prepopulated. I'm just not sure that it's always the same path. But I'd need to find out.

|===
+
. Click *Create Project*.
+
The flow builder canvas appears with the following starting cards: 

image::int-gen-flow-starting-cards.png[Starting cards on new project canvas]
[callout list]

> . Click *Open Einsten* in the *Start with AI* card to use the generative flow feature.

Reviewer: ```suggestion
. Click *Open Einstein* in the *Start with AI* card to use the generative flow feature.
```

. Click *Build a flow* in the *Start From Scratch* card to begin building your flow from scratch.

[[craft-ai-prompt]]
== Craft Your Prompt

> Before you create a flow, define what you want achieve with the flow, for example, optimizing a business process. Craft your prompt using natural language, and keep in mind that the more concise and specific you make your prompt, the more effective it will be when it generates your flow. 

Reviewer: ```suggestion
Before you create a flow, define what you want achieve with the flow, for example, optimizing a business process. Craft your prompt using natural language, and keep in mind that the more concise and specific you make your prompt, the more effective Einstein will be when generating your flow. 
```
or something like that

Here are some best practices for writing prompts:

* Be clear and precise:
+ 
Use clear and specific language to describe the use case for the flow to build.
* Include trigger events: 

> +

Reviewer: re: "trigger events" maybe "triggering events"?

Mention the trigger events that initiate the flow, for example, sending an email when a new case is created. 
* Emphasize key components:
+

> Highlight the key MuleSoft components to include in the flow, for example, HTTP *Listener* and Email *Send*.

Reviewer: maybe "components and connector operations"

Reviewer: as I mention later, we need to decide how to sync on our language

* Specify data objects:

> +

Reviewer: Should the next bullet ("* Clearly mention...") be a bullet?

> * Clearly mention the data objects to include in the flow, such as "Salesforce Account object" or any other relevant data source.

Reviewer: ```suggestion
** Clearly mention the data objects to include in the flow, such as "Salesforce Account object" or any other relevant data source.
```

* Proofread and revise:
+
Before submitting the prompt, proofread your prompt to ensure that it is clear and comprehensive.

> 

Reviewer: Do we have examples we could provide, or are they in the product?

Reviewer: Oh, I see on the next page. Could we provide a link to that info? I felt like I wanted an example or two when I was reading these guidelines.

Author: @dukesphere There are example prompts in the example topic. Do you think we should link to that topic here, or...?

Reviewer: I was thinking a link in case someone want something concrete to look at.

+
image::int-generated-case-notification-email.png[]
+
. Optionally, click the thumbs up or thumbs down icon that appears after the generated code. 
+
//I made this up. Does this work this way?
Rating the flow that's created for you helps Einstein learn so that over time, the generated flows improve.
. Inspect the generated code, and if you are satisfied, click the insert code icon to insert the code in to your flow. 
+
If you aren't satisfied with the generated flow, reword your prompt and send it again. 

[[customize-gen-flow]]
== Complete Your Flow

Generative Flow builds the initial flow for you so that all you need to do is fill in the information that's specific to your configuration. 

. After you inspect the code for your generated flow, click the Insert icon. 
+
After the code is inserted, the initial flow is built and the visual representation of the flow is displayed in the flow canvas next to the XML code tab. 
+
The error icon is displayed for components that require further configuration. 
. To determine where the error is, select the component in the canvas.
+
Anypoint Code Builder highlights the location of the error within the configuration XML.

> . 

Reviewer: ```suggestion

```
stray period, or maybe another step is coming

= Creating Integrations Using AI

== Before You Begin

Before you start creating your integration using the generative AI feature:

* Ensure that you have the following permissions:
** Anypoint Code Builder Developer
** Mule Developer Generative AI User
+
If you see the *Start with AI* card in the canvas, you have the correct permissions.
* Install a REST client such as Postman or Advanced REST Client. 

[[create-integration-project]]
== Create a New Integration Project

To create a new integration project:

// Open the ACB IDE
include::partial$acb-reusable-steps.adoc[tags="open-ide"]
. From *Quick Actions*, click *Develop an Integration*:
+
image::develop-integration-option-mat.png["Develop an Integration link highlighted in the Getting started section"]

. Complete the *Develop an Integration* form.
+
[%header,cols="20a,60a"] 
|===
| Field Name | Field Value

| *Project Name* | Unique name for your project.

This name is used as the title and name of the integration specification file.

> For example, if the project name is "OAS Example", the specification file name is `oas-example`.

Reviewer: Let's use something other than OAS Example (since that's an API spec language) 

Could use "My Integration" or something like that. Adding a suggestion.

> For example, if the project name is "OAS Example", the specification file name is `oas-example`.

Reviewer: ```suggestion
For example, if the project name is "My Integration", the name of the folder for the integration project is my-integration-app, and the configuration XML file name is `my-integration-app.xml`.
```

Not sure we need all that detail, but you get the idea.

| *Project Location* | Your home directory or another directory you create. Click *Browse* to select a different directory for the integration.
|===
+
. Click *Create Project*.
+
The flow builder canvas appears with two cards:

//Is that called the flow builder canvas or...? Should we have an image that shows the two cards?

You can:

* Click *Open Einsten* in the *Start with AI* card
* 

== Add a Flow to Your Project

You can build flows using generative AI or you can develop the flow on your own, from scratch. 

To use generative AI:

> . If you already have a good idea of the type of flow you want to design, click *Open Einsten* in the *Start with AI* card.

Reviewer: ```suggestion
. If you already have a good idea of the type of flow you want to design, click *Open Einstein* in the *Start with AI* card.
```
